### PR TITLE
#35711: Fix Smooth Scrolling on Linux with Display Scaling higher than 100%

### LINKED ISSUE #35711: Smooth Scrolling not working on HiDPI displays if Display Scaling is higher than 100%
- VSCode Version: Code 1.17.0 (be377c0faf7574a59f84940f593a6849f12e4de7, 2017-10-04T23:45:46.804Z)
- OS Version: Windows_NT x64 10.0.15063
- Extensions: none

---

Steps to Reproduce:

1. set "editor.smoothScrolling": true
2. close and re-open VS Code
3. try scrolling the editor - it will jump as if "editor.smoothScrolling" were set to False

Reproduces without extensions: Yes (tried to launch with _code --disable-extensions_, same issue)

To be honest I don't know whether this is an issue related to this specific version as it's the first time I'm using VS Code.

My mouse is an MX Master and I'm using an ASUS N550 laptop.
Anybody else experiencing this issue?

Thanks

### PR DESCRIPTION
**Pull request by krlvm**
This PR fixes #35711

### CODE REVIEW

**Comment by krlvm** [comment]
It should work on macOS too (I can't check), so I did not add platform check.
Otherwise, I had to add `platform.isLinux` check

**Comment by krlvm** [comment]
More accurate method:
`classifier.accept(Date.now(), Math.floor(Math.ceil(e.deltaX) / (Math.round(window.devicePixelRatio * 100) / 100)), Math.floor(Math.ceil(e.deltaY) / (Math.round(window.devicePixelRatio * 100) / 100)));`

File: src/vs/base/browser/ui/scrollbar/scrollableElement.ts [hunk 1]
```diff
@@ -348,7 +348,12 @@ export abstract class AbstractScrollableElement extends Widget {
 				// so to get a better classification, simply undo that.
 				classifier.accept(Date.now(), e.deltaX / window.devicePixelRatio, e.deltaY / window.devicePixelRatio);
 			} else {
-				classifier.accept(Date.now(), e.deltaX, e.deltaY);
+				if(platform.isLinux) {
```
Comment (ItzLevvie) [inline src/vs/base/browser/ui/scrollbar/scrollableElement.ts:351]:
```suggestion
				if (platform.isLinux) {
```

File: src/vs/base/browser/ui/scrollbar/scrollableElement.ts [hunk 2]
```diff
@@ -348,7 +348,12 @@ export abstract class AbstractScrollableElement extends Widget {
 				// so to get a better classification, simply undo that.
 				classifier.accept(Date.now(), e.deltaX / window.devicePixelRatio, e.deltaY / window.devicePixelRatio);
 			} else {
-				classifier.accept(Date.now(), e.deltaX, e.deltaY);
+				if (platform.isLinux) {
+					// There's a problem with floating point on Linux
+					classifier.accept(Date.now(), Math.ceil(e.deltaX/window.devicePixelRatio), Math.ceil(e.deltaY/window.devicePixelRatio));
```
Comment (ItzLevvie) [inline src/vs/base/browser/ui/scrollbar/scrollableElement.ts:353]:
```suggestion
					classifier.accept(Date.now(), Math.ceil(e.deltaX / window.devicePixelRatio), Math.ceil(e.deltaY / window.devicePixelRatio));
```

**Review by alexdima — CHANGES_REQUESTED** [review]
The classifier of physical mouse wheels uses two signals: 
* if both `deltaX` and `deltaY` are non zero this indicates that the event came from a touchpad.
* if the `deltaX`/`deltaY` values are integers, this indicates that the event came from a mouse wheel.

This proposed change always turns the values into integers, so only the first signal remains on Linux. We can hard-code that the integer signal is not valid on Linux, but then I would prefer a more explicit check inside the classifier itself rather than in the call site as proposed here.

**Comment by krlvm** [comment]
I can suggest another temporary fix based on statistics saying that the number is always in the range 52.9 to 53 that will not break touchpad check.

**Pull request closed without merging**